### DISCUSSION TITLE
farZOffset instead of farZMultiplier (underwater data)

### DISCUSSION THREAD

**Discussion by jlumsden-mts** [visgl/deck.gl#7754] [Q&A] [answered]
Hello, we have a use case where we are rendering underwater data below sea level. As per the recommendations we have increased the farZMultiplier to 100 or even 1,000,000 but we still have issues with farZ clipping at high zoom levels as the camera approaches sea level.

I assume this is due to the camera height above sea level calculations here:
https://github.com/uber-web/math.gl/blob/d618c26869a15f277af56df8416216cf07e5c747/modules/web-mercator/src/web-mercator-utils.ts#L316

What I really want to do is add a fixed offset to farZ given I know that the data will be no more than 11km below sea level (maximum ocean depth) but I can't see a way to do this. There seems to be a mechanism for supplying a custom projection matrix to MapView but I've not been able to get this to work. Any ideas please?

**Comment by Pessimistress** [comment, marked as answer]
I don't think there is a way to arbitrarily set farZ at the moment, thought I can see why it would be useful in your case. Right now `farZMultiplier` is still limited by another variable, so simply blowing it up does not do what you expect:

https://github.com/uber-web/math.gl/blob/d618c26869a15f277af56df8416216cf07e5c747/modules/web-mercator/src/web-mercator-utils.ts#L334

I think the right way to apply `farZMultiplier` should be

```js
const farZ = Math.min(furthestDistance, horizonDistance) * farZMultiplier;
```

Note that where farZ maps to the real-world coordinate is dependent on the zoom level, so you may still have to adjust it as the zoom level change.

**Reply by Pessimistress** [reply]
Opened https://github.com/uber-web/math.gl/pull/330

**Reply by jlumsden-mts** [reply]
Thanks for your response! The change you suggested in math.gl makes sense to me.

I did have some success yesterday generating a custom projectionMatrix and adjusting the center[2] coordinate from the default 0 to a larger number in order to increase cameraToSeaLevelDistance:

https://github.com/uber-web/math.gl/blob/d618c26869a15f277af56df8416216cf07e5c747/modules/web-mercator/src/web-mercator-utils.ts#L315

However it was quite messy overriding viewState in onViewStateChange so your suggested change is definitely preferable noting that, as you mentioned, we may still have to adjust it for scale / zoom in order to achieve the right draw distance.

**Comment by jlumsden-mts** [comment]
Thanks for the assistance
